packages/flutter/lib/src/material/text_field.dart

```diff
@@ -418,10 +460,23 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
         hintMaxLines: widget.decoration?.hintMaxLines ?? widget.maxLines
       );
 
+    Widget counter;
```

Comment:
The internal logic for this method should be shuffled around a bit to make what's going on clearer.

Probably init the needsCounter var right before the `if (!needsCounter)` short-circuit return and include the `buildCounter != null` case too. And then use buildCounter or the counterText code...